src/librustc/mir/interpret/mod.rs

```diff
@@ -199,23 +227,48 @@ pub fn specialized_encode_alloc_id<
 
 pub fn specialized_decode_alloc_id<
     'a, 'tcx,
-    D: Decoder,
-    CACHE: FnOnce(&mut D, AllocId),
+    D: TyDecoder<'a, 'tcx>,
+    GlobalCache: FnMut(&mut D) -> LockGuard<'_, FxHashMap<usize, (AllocId, bool)>>,
+    LocalCache: FnOnce(&mut D) -> &mut FxHashSet<AllocId>,
 >(
     decoder: &mut D,
     tcx: TyCtxt<'a, 'tcx, 'tcx>,
-    cache: CACHE,
+    mut global_cache: GlobalCache,
+    local_cache: LocalCache,
 ) -> Result<AllocId, D::Error> {
+    let pos = decoder.position();
     match AllocKind::decode(decoder)? {
+        AllocKind::AllocAtPos => {
```

Comment:
You could use the trick here that I had originally where you read a usize, and that tag is either 0 for `Alloc`, 1 for `Static`, 2 for `Fn` or anything else is the `real_pos`. This is also used in `Ty` encoding I think.